include/deal.II/base/tensor.h

```diff
@@ -19,741 +19,213 @@
-  return ( Tensor<rank_-1,dim,Number>::component_to_unrolled_index(indices1) * dim + indices[rank_-1]);
-}
+//@}
+/**
+ * @name Vector space operations on Tensor objects:
+ */
+//@{
 
-template <int rank_, int dim, typename Number>
+/**
+ * Scalar multiplication of a tensor of rank 0 with a scalar from the left.
+ *
+ * @relates Tensor<0,dim,Number>
+ * @relates EnableIfScalar
+ */
+template <int dim,
+          typename Number,
+          typename OtherNumber,
+          typename = typename EnableIfScalar<OtherNumber>::type>
 inline
-TableIndices<rank_>
-Tensor<rank_, dim, Number>::unrolled_to_component_indices(const unsigned int i)
+Tensor<0,dim,typename ProductType<OtherNumber, Number>::type>
+operator * (const OtherNumber           factor,
+            const Tensor<0,dim,Number> &t)
 {
-  Assert (i < n_independent_components,
-          ExcIndexRange (i, 0, n_independent_components));
-
-  TableIndices<rank_>   indices;
-
-  unsigned int remainder = i;
-  for (int r=rank_-1; r>=0; --r)
-    {
-      indices[r] = (remainder % dim);
-      remainder /= dim;
-    }
-  Assert (remainder == 0, ExcInternalError());
-
-  return indices;
+  return factor * static_cast<Number>(t);
 }
 
 
-
-template <int rank_, int dim, typename Number>
+/**
+ * Scalar multiplication of a tensor of rank 0 with a scalar from the right.
+ *
+ * @relates Tensor<0,dim,Number>
+ * @relates EnableIfScalar
+ */
+template <int dim,
+          typename Number,
+          typename OtherNumber,
+          typename = typename EnableIfScalar<OtherNumber>::type>
 inline
-void Tensor<rank_,dim,Number>::clear ()
+Tensor<0,dim,typename ProductType<Number, OtherNumber>::type>
+operator * (const Tensor<0,dim,Number> &t,
+            const OtherNumber           factor)
 {
-  for (unsigned int i=0; i<dim; ++i)
-    subtensor[i].clear();
+  return static_cast<Number>(t) * factor;
 }
 
 
-
-template <int rank_, int dim, typename Number>
+/**
+ * Division of a tensor of rank 0 with a scalar number.
+ *
+ * @relates Tensor<0,dim,Number>
+ * @relates EnableIfScalar
+ */
+template <int dim,
+          typename Number,
+          typename OtherNumber,
+          typename = typename EnableIfScalar<OtherNumber>::type>
 inline
-std::size_t
-Tensor<rank_,dim,Number>::memory_consumption ()
+Tensor<0,dim,typename ProductType<Number, OtherNumber>::type>
+operator / (const Tensor<0,dim,Number> &t,
+            const OtherNumber           factor)
 {
-  return sizeof(Tensor<rank_,dim,Number>);
+  return static_cast<Number>(t) / factor;
 }
 
 
-
-template <int rank_, int dim, typename Number>
-template <class Archive>
+/**
+ * Add two tensors of rank 0.
+ *
+ * @relates Tensor<0,dim,Number>
+ */
+template <int dim, typename Number, typename OtherNumber>
 inline
-void
-Tensor<rank_,dim,Number>::serialize(Archive &ar, const unsigned int)
+Tensor<0, dim, typename ProductType<Number, OtherNumber>::type>
+operator+ (const Tensor<0,dim,Number> &p, const Tensor<0,dim,OtherNumber> &q)
```

Comment:
So this is a good question -- does `ProductType` hold the appropriate data type for an addition?